design-documents/remote-invoker.md

```diff
@@ -0,0 +1,618 @@
+    public function createStarted(string $callId, int $timeout): void;
+    
+    public function saveResponse(string $callId, InvokeResponseInterface $response): void;
+    
+    /**
+     * @throws NoSuchEntityException When there were no such call registered.
+     * @return InvokeResponseInterface|null Null if call is not yet finished.
+     */
+    public function get(string $callId): ?InvokeResponseInterface;
+}
+```
+##### Remote call processor
+Perform remote call
+```
+interface LocalInvokerInterface {
+    public function invoke(InvokeRequestInterface $request): InvokeResponseInterface;
+}
+```
+##### Response generator
+Transform invoke response to a regular response
+```
+interface ResponseGeneratorInterface {
+    public function generate(string $callId, InvokeResponseInterface $response): ResponseInterface;
+}
+```
+
+
+### Networking gateway
+
+_Remote code invoker_ is an abstraction, it's not enforcing how requests and responses are actually sent and parsed.
+
+#### RPC-style gateway
+
+HTTP protocol will be used. A controller will be added to process a single URL like _magento-base-url_/invoker.
+Requests' body will contain requested class name, method name, arguments and it all encoded in json.
+Requests' headers will contain information such as authenticated user ID, call ID, issued timestamp, signature.
+Response body will contain serialized service call result or a serialized thrown exception encoded in json.
+
+#### Why RESTful web API should not be used as the network gateway for the invoker
+* Not all service contracts are exposed as web API and nor should they be (for security and workflow reasons)
+_([see explanation](#exposing-internal-logic-as-an-external-web-api))_
+* Web API methods have certain ACL resources assigned to them so the ACL checks would be
+performed every time a proxy service calls remote service unlike when a service
+is called locally _([see explanation](#redundant-acl-validations-when-using-existing-web-api))_
+* Proxy services would have to generate auth information of already authorized users (tokens etc) to access REST/SOAP
+services which is a redundant process, also some of those REST/SOAP endpoints would have limitations in
+place for certain type of users which would be useless in case of remote execution
+(like requiring an admin user) _([see explanation](#redundant-authorization-validation-when-using-existing-web-api))_
+* Other application state spoofing may be required for seamless remote execution of services
+* RESTful Web API provides access to service contracts by assigning a unique human-readable
+HTTP method/URL pair for external systems - we don't need that when writing a proxy service since we
+would always already have class and method's names, the only other thing existing RESTful Web API mechanism
+provides is it's serializer and it will be reused in the _Remote Code Invoker_
+* When an exception occurs during a service call via RESTful API the response does not contain
```

Comment:
Looks more like an issue with error reporting for Web APIs.

Reply:
not really, clientside and 3rd-party systems wouldn't care for exception classes in responses of RESTful APIs